// @candidosales/material-time-picker/esm2015/lib/w-time-dialog/w-time-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { CLOCK_TYPE } from '../w-clock/w-clock.component';
import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA } from '@angular/material';
import { MatDialogRef } from '@angular/material';
export class WTimeDialogComponent {
    /**
     * @param {?} dialogRef
     * @param {?} data
     */
    constructor(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.VIEW_HOURS = CLOCK_TYPE.HOURS;
        this.VIEW_MINUTES = CLOCK_TYPE.MINUTES;
        this.currentView = this.VIEW_HOURS;
        this.userTime = this.data.time;
        this.color = this.data.color;
    }
    /**
     * @return {?}
     */
    revert() {
        this.dialogRef.close(-1);
    }
    /**
     * @return {?}
     */
    submit() {
        this.dialogRef.close(this.userTime);
    }
}
WTimeDialogComponent.decorators = [
    { type: Component, args: [{
                template: "\n<div mat-dialog-content class=\"w-timepicker-dialog\">\n    <w-time [color]=\"color\" [userTime]=\"userTime\" (onRevert)=\"revert()\" (onSubmit)=\"submit()\" [revertLabel]=\"data.revertLabel\" [submitLabel]=\"data.submitLabel\"></w-time>\n</div>\n\n",
                styles: [".w-timepicker-dialog{padding:0;margin:-24px;width:calc(100% + 48px)}"]
            }] }
];
/** @nocollapse */
WTimeDialogComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
if (false) {
    /** @type {?} */
    WTimeDialogComponent.prototype.userTime;
    /** @type {?} */
    WTimeDialogComponent.prototype.color;
    /** @type {?} */
    WTimeDialogComponent.prototype.VIEW_HOURS;
    /** @type {?} */
    WTimeDialogComponent.prototype.VIEW_MINUTES;
    /** @type {?} */
    WTimeDialogComponent.prototype.currentView;
    /** @type {?} */
    WTimeDialogComponent.prototype.dialogRef;
    /** @type {?} */
    WTimeDialogComponent.prototype.data;
}
