// @candidosales/material-time-picker/esm2015/lib/material-timepicker/material-timepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { MatDialog } from '@angular/material';
import { WTimeDialogComponent } from '../w-time-dialog/w-time-dialog.component';
import { Utils } from '../utils';
export class MaterialTimePickerComponent {
    /**
     * @param {?} dialog
     */
    constructor(dialog) {
        this.dialog = dialog;
        this.label = 'Hour';
        this.appearance = 'legacy';
        this.change = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.userTime) {
            this.userTime = {
                hour: 0,
                minute: 0,
                meriden: 'PM',
                format: 24
            };
        }
    }
    /**
     * @return {?}
     */
    time() {
        if (!this.userTime) {
            return '';
        }
        /** @type {?} */
        let meriden = `${this.userTime.meriden}`;
        if (this.userTime.format === 24) {
            meriden = '';
        }
        /** @type {?} */
        let hour = `${this.userTime.hour}`;
        if (this.userTime.hour === 24) {
            hour = '00';
        }
        if (this.userTime.minute === 0) {
            return `${hour}:00 ${meriden}`;
        }
        else if (this.userTime.minute < 10) {
            /** @type {?} */
            const tt = '0' + String(this.userTime.minute);
            return `${hour}:${tt} ${meriden}`;
        }
        else {
            return `${hour}:${this.userTime.minute} ${meriden}`;
        }
    }
    /**
     * @return {?}
     */
    showPicker() {
        /** @type {?} */
        const dialogRef = this.dialog.open(WTimeDialogComponent, {
            data: {
                time: {
                    hour: this.userTime.hour,
                    minute: this.userTime.minute,
                    meriden: this.userTime.meriden,
                    format: this.userTime.format
                },
                color: this.color,
                revertLabel: this.revertLabel,
                submitLabel: this.submitLabel
            }
        });
        dialogRef.afterClosed().subscribe((result) => {
            // result will be update userTime object or -1 or undefined (closed dialog w/o clicking cancel)
            if (result === undefined) {
                return;
            }
            else if (result !== -1) {
                this.userTime = result;
                /** @type {?} */
                const hour = result.hour;
                /** @type {?} */
                const minute = result.minute;
                /** @type {?} */
                const dataEvent = {
                    hour: Utils.formatHour(result.format, hour),
                    minute: Utils.formatMinute(minute),
                    meriden: this.userTime.meriden,
                    format: this.userTime.format
                };
                this.emitChange(dataEvent);
            }
        });
        return false;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    emitChange(data) {
        this.change.emit(data);
    }
}
MaterialTimePickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'material-timepicker',
                template: "<div fxFlex fxLayout=\"row\" class=\"w-mat-timepicker\">\n  <mat-form-field fxFlex class=\"timeContainer\" appearance=\"{{appearance}}\">\n      <mat-label>{{ label }}</mat-label>\n      <input matInput class=\"timeInput\" placeholder=\"Selec time\" id=\"time_Control\" name=\"time_Control\" [value]=\"time()\">\n      <mat-icon matSuffix (click)=\"showPicker()\">access_time</mat-icon>\n  </mat-form-field>\n</div>\n",
                styles: [".time-picker-button{padding:0;margin:0;min-width:44px}"]
            }] }
];
/** @nocollapse */
MaterialTimePickerComponent.ctorParameters = () => [
    { type: MatDialog }
];
MaterialTimePickerComponent.propDecorators = {
    label: [{ type: Input }],
    appearance: [{ type: Input }],
    userTime: [{ type: Input }],
    color: [{ type: Input }],
    revertLabel: [{ type: Input }],
    submitLabel: [{ type: Input }],
    change: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    MaterialTimePickerComponent.prototype.label;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.appearance;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.userTime;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.color;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.revertLabel;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.submitLabel;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.change;
    /** @type {?} */
    MaterialTimePickerComponent.prototype.dialog;
}
