// @candidosales/material-time-picker/esm5/lib/w-time/w-time.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { CLOCK_TYPE } from '../w-clock/w-clock.component';
import { Utils } from '../utils';
import { Component, EventEmitter, Input, Output } from '@angular/core';
var WTimeComponent = /** @class */ (function () {
    function WTimeComponent() {
        this.userTimeChange = new EventEmitter();
        this.onRevert = new EventEmitter();
        this.onSubmit = new EventEmitter();
        this.VIEW_HOURS = CLOCK_TYPE.HOURS;
        this.VIEW_MINUTES = CLOCK_TYPE.MINUTES;
        this.currentView = this.VIEW_HOURS;
    }
    /**
     * @return {?}
     */
    WTimeComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.userTime) {
            this.userTime = {
                hour: 6,
                minute: 0,
                meriden: 'PM',
                format: 12
            };
        }
        if (!this.revertLabel) {
            this.revertLabel = 'Cancel';
        }
        if (!this.submitLabel) {
            this.submitLabel = 'Okay';
        }
    };
    /**
     * @param {?} format
     * @param {?} hour
     * @return {?}
     */
    WTimeComponent.prototype.formatHour = /**
     * @param {?} format
     * @param {?} hour
     * @return {?}
     */
    function (format, hour) {
        return Utils.formatHour(format, hour);
    };
    /**
     * @param {?} minute
     * @return {?}
     */
    WTimeComponent.prototype.formatMinute = /**
     * @param {?} minute
     * @return {?}
     */
    function (minute) {
        return Utils.formatMinute(minute);
    };
    /**
     * @param {?} type
     * @return {?}
     */
    WTimeComponent.prototype.setCurrentView = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        this.currentView = type;
    };
    /**
     * @param {?} m
     * @return {?}
     */
    WTimeComponent.prototype.setMeridien = /**
     * @param {?} m
     * @return {?}
     */
    function (m) {
        this.userTime.meriden = m;
    };
    /**
     * @return {?}
     */
    WTimeComponent.prototype.revert = /**
     * @return {?}
     */
    function () {
        this.onRevert.emit();
    };
    /**
     * @return {?}
     */
    WTimeComponent.prototype.submit = /**
     * @return {?}
     */
    function () {
        this.onSubmit.emit(this.userTime);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    WTimeComponent.prototype.emituserTimeChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.userTimeChange.emit(this.userTime);
    };
    WTimeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'w-time',
                    template: "<div fxLayout=\"row\" fxLayout.lt-md=\"column\" fxLayoutAlign=\"center center\" class=\"w-time\" [ngClass.xs]=\"'vertical-time'\" [ngClass.sm]=\"'vertical-time'\">\n  <mat-toolbar fxLayout=\"column\"  fxLayout.lt-md=\"row\" fxLayoutAlign=\"center center\" [color]=\"color\" class=\"w-timepicker-time-container\">\n\n      <div class=\"w-timepicker-selected-time\">\n          <span [class]=\"currentView === VIEW_HOURS ? 'active': ''\" (click)=\"setCurrentView(VIEW_HOURS)\">{{ formatHour(userTime.format, userTime.hour) }}:</span>\n          <span [class]=\"currentView === VIEW_MINUTES ? 'active': ''\" (click)=\"setCurrentView(VIEW_MINUTES)\">{{ formatMinute(userTime.minute) }}</span>\n      </div>\n      <div fxLayout=\"column\" fxLayoutAlign=\"center center\" class=\"w-timepicker-selected-ampm\" *ngIf=\"userTime.format === 12\">\n          <span (click)=\"setMeridien('AM')\" [class]=\"userTime.meriden === 'AM' ? 'active' : ''\">AM</span>\n          <span (click)=\"setMeridien('PM')\" [class]=\"userTime.meriden === 'PM' ? 'active' : ''\">PM</span>\n      </div>\n\n  </mat-toolbar>\n\n  <div fxLayout=\"column\" fxLayoutAlign=\"space-between center\" class=\"w-time-content\">\n      <w-clock [color]=\"color\" class=\"w-animation-zoom\" [userTime]=\"userTime\" (userTimeChange)=\"emituserTimeChange($event)\" [(currentView)]=\"currentView\" (viewChange)=\"setCurrentView($event)\"></w-clock>\n\n      <div fxFlexAlign=\"end\">\n          <button mat-button (click)=\"revert()\">{{revertLabel}}</button>\n          <button mat-button [color]=\"color\" (click)=\"submit()\">{{submitLabel}}</button>\n      </div>\n  </div>\n</div>\n",
                    styles: [":host{display:block}.w-time{max-height:100%;min-height:350px;height:350px}.w-time .w-timepicker-time-container{padding:15px;min-width:160px;width:160px}.w-time .w-timepicker-time-container.mat-toolbar-single-row{height:100%}.w-time .w-timepicker-selected-time{font-size:3.5rem;font-weight:400;display:flex}.w-time .w-timepicker-selected-ampm{font-size:1rem;line-height:1.3rem;padding-top:2rem}.w-time .w-time-content{width:100%;height:100%;padding:6px}.w-time .w-time-content w-clock{padding:12px 0;height:calc(100% - 58px)}.w-time.vertical-time{height:auto}.w-time.vertical-time .w-timepicker-time-container{min-width:auto;width:100%;height:100px}.w-time.vertical-time .w-timepicker-time-container .w-timepicker-selected-ampm{padding:0 12px}.w-timepicker-selected-ampm>span,.w-timepicker-selected-time>span{outline:0;opacity:.5}.w-timepicker-selected-ampm>span:not(.active),.w-timepicker-selected-time>span:not(.active){cursor:pointer}.w-timepicker-selected-ampm>span.active,.w-timepicker-selected-time>span.active{opacity:1}.w-animate-next{opacity:0;-webkit-transform:translate3d(50%,0,1px);transform:translate3d(50%,0,1px)}.w-animate-next-remove{transition:.5s cubic-bezier(.35,0,.25,1);opacity:0;-webkit-transform:translate3d(50%,0,1px);transform:translate3d(50%,0,1px)}.w-animate-next-remove-active{opacity:1;-webkit-transform:translate3d(0,0,1px);transform:translate3d(0,0,1px)}.w-animate-prev{opacity:0;-webkit-transform:translate3d(-50%,0,1px);transform:translate3d(-50%,0,1px)}.w-animate-prev-remove{transition:.3s cubic-bezier(.35,0,.25,1);opacity:0;-webkit-transform:translate3d(-50%,0,1px);transform:translate3d(-50%,0,1px)}.w-animate-prev-remove-active{opacity:1;-webkit-transform:translate3d(0,0,1px);transform:translate3d(0,0,1px)}@-webkit-keyframes w-animation-bounce{from{opacity:0;-webkit-transform:scale(.95);transform:scale(.95)}70%{opacity:1;-webkit-transform:scale(1.05);transform:scale(1.05)}to{-webkit-transform:scale(1);transform:scale(1)}}@keyframes w-animation-bounce{from{opacity:0;-webkit-transform:scale(.95);transform:scale(.95)}70%{opacity:1;-webkit-transform:scale(1.05);transform:scale(1.05)}to{-webkit-transform:scale(1);transform:scale(1)}}.w-animation-zoom.ng-enter{transition:.3s cubic-bezier(.35,0,.25,1);-webkit-animation-duration:.3s;animation-duration:.3s;-webkit-animation-name:w-animation-bounce;animation-name:w-animation-bounce}"]
                }] }
    ];
    /** @nocollapse */
    WTimeComponent.ctorParameters = function () { return []; };
    WTimeComponent.propDecorators = {
        userTime: [{ type: Input }],
        userTimeChange: [{ type: Output }],
        revertLabel: [{ type: Input }],
        submitLabel: [{ type: Input }],
        onRevert: [{ type: Output }],
        onSubmit: [{ type: Output }],
        color: [{ type: Input }]
    };
    return WTimeComponent;
}());
export { WTimeComponent };
if (false) {
    /** @type {?} */
    WTimeComponent.prototype.userTime;
    /** @type {?} */
    WTimeComponent.prototype.userTimeChange;
    /** @type {?} */
    WTimeComponent.prototype.revertLabel;
    /** @type {?} */
    WTimeComponent.prototype.submitLabel;
    /** @type {?} */
    WTimeComponent.prototype.onRevert;
    /** @type {?} */
    WTimeComponent.prototype.onSubmit;
    /** @type {?} */
    WTimeComponent.prototype.color;
    /** @type {?} */
    WTimeComponent.prototype.VIEW_HOURS;
    /** @type {?} */
    WTimeComponent.prototype.VIEW_MINUTES;
    /** @type {?} */
    WTimeComponent.prototype.currentView;
}
